cleverhans/utils.py

```diff
@@ -220,6 +220,101 @@ def grid_visual(data):
     return figure
 
 
+def get_logits_over_interval(sess, model, x, adv_x, x_data,
+                             min_epsilon=-10, max_epsilon=10,
+                             num_points=21):
+    """Get logits when the input is perturbed in an interval in adv direction.
+
+    Args:
+        sess: Tf session
+        model: Model for which we wish to get logits
+        x: Data tensor of shape [None, height, width, channels]
+        adv_x: Tensor for adversarial perturbation of x
+        x_data: Numpy array corresponding to single data
+                point of shape [height, width, channels].
+        min_epsilon: Minimum value of epsilon over the interval
+        max_epsilon: Maximum value of epsilon over the interval
+        num_points: Number of points used to interpolate
+
+    Returns:
+        Numpy array containing logits
+
+    Raises:
+        ValueError if min_epsilon is larger than max_epsilon
+    """
+    import tensorflow as tf
+    if min_epsilon > max_epsilon:
+        raise ValueError('Minimum epsilon is less than maximum epsilon')
+
+    eta = tf.nn.l2_normalize(adv_x - x, dim=0)
```

Comment:
You should be able to keep this in TensorFlow rather than NumPy. Also, you should be able to do everything using broadcasting, rather than building a larger array.

Something like

     epsilon = tf.reshape(tf.lin_space(min_epsilon, max_epsilon, num_points), (num_points, 1, 1, 1))
     lin_batch = x + epsilon * eta

should be able to do it without any explicit duplication of x and eta.

Reply:
Done